### PR TITLE
add heartbeat between worker and log_server

### PR DESCRIPTION
**Pull request by zenghsh3**

### CODE REVIEW

File: parl/remote/heartbeat_server.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/heartbeat_server.py]: Please update the year in copyright.

File: parl/remote/heartbeat_server.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class HeartbeatServerThread(threading.Thread):
+    def __init__(self, heartbeat_client_recv_addr, server_type, client_type):
+        """Create a thread to run the heartbeat server.
+
+        Args:
+            heartbeat_client_recv_addr(str): address of another server socket in the heartbeat client 
+                    which used to receive the heartbeat server address.
+            server_type(str): type of server, which is used for logging.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/heartbeat_server.py]: Please tell the developer the possible types for this argument.
Check if the passed type is one of the possible types.
Same suggestion for the `client_type` argument.

File: parl/remote/heartbeat_server.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import threading
+import zmq
+from parl.remote import remote_constants
+from parl.remote.zmq_utils import create_server_socket, create_client_socket
+from parl.utils import get_ip_address, to_byte, logger
+
+
+class HeartbeatServerThread(threading.Thread):
+    def __init__(self, heartbeat_client_recv_addr, server_type, client_type):
+        """Create a thread to run the heartbeat server.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/heartbeat_server.py:25]: add the comment:
Note that the process will exit directly if the heartbeat detection fails.

File: parl/remote/heartbeat_server.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/heartbeat_server.py:2]: Can we add some tests for this file?

File: parl/remote/worker.py [hunk 5]
```diff
@@ -121,11 +122,12 @@ def check_version(self):
             raise NotImplementedError
 
     def _create_sockets(self):
-        """ Each worker has three sockets at start:
+        """ Each worker has four sockets at start:
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/worker.py]: Each worker maintains four sockets:

File: parl/remote/worker.py [hunk 6]
```diff
@@ -410,8 +426,45 @@ def _create_log_server(self, port):
         FNULL.close()
 
         log_server_address = "{}:{}".format(self.worker_ip, port)
+
+        message = self.reply_log_server_socket.recv_multipart()
+        log_server_heartbeat_addr = to_str(message[1])
+        self.reply_log_server_socket.send_multipart(
+            [remote_constants.NORMAL_TAG])
+
+        # a thread for sending heartbeat signals to job
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/worker.py]: It sends the signal to job?

File: parl/remote/heartbeat_server.py (hunk 4 above)
Comment (zenghsh3) [reply parl/remote/heartbeat_server.py:2]: We will refactor the code of the heartbeat mechanism recently, so we will add the unitest after the refactoring.

File: parl/remote/heartbeat_server.py (hunk 1 above)
Comment (zenghsh3) [reply parl/remote/heartbeat_server.py]: done

File: parl/remote/heartbeat_server.py (hunk 2 above)
Comment (zenghsh3) [reply parl/remote/heartbeat_server.py]: done

File: parl/remote/heartbeat_server.py (hunk 3 above)
Comment (zenghsh3) [reply parl/remote/heartbeat_server.py:25]: done

File: parl/remote/worker.py (hunk 5 above)
Comment (zenghsh3) [reply parl/remote/worker.py]: done

File: parl/remote/worker.py (hunk 6 above)
Comment (zenghsh3) [reply parl/remote/worker.py]: changed to `log_server`

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Pull request merged**